### ISSUE TITLE
The Rete Algorithm Explained!

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#3038]
The Rete Algorithm Explained!<br>
<br>
https://ift.tt/2pXHFFO<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div id="post-1196">
<p>In part 1 of this blog series, I covered <a title="Rete Algorithm Demystified! &mdash; Part&nbsp;1" href="https://www.sparklinglogic.com/rete-algorithm-demystified-part-1/">the Rete Algorithm and its origin</a> and even <a title="Rete Algorithm Demystified! &mdash; Part 1&nbsp;Answer!" href="https://www.sparklinglogic.com/rete-algorithm-demystified-part-1-answer/">the origin of its name</a>.&nbsp; Now&nbsp;as promised, I am going to explain how it works.&nbsp; It is a challenge to keep it simple not to lose our less technical audience and yet interesting enough for the techies.&nbsp; I hope I will reach that goal.&nbsp; I am trying to convince Carlos to do a Part 3 that will speak to the hard-core techies on more advanced considerations.</p>
<p>And without further due, let me introduce the Rete algorithm!</p>
<p><strong>How it works</strong></p>
<p>The brilliant idea that Dr. Charles Forgy developed was to decouple the evaluation of hypothesis from execution sequencing.</p>
<p>Why is that brilliant you might ask?&nbsp; When large volumes of business rules need to be assessed against facts or data, the constant screening for applicability can be costly &mdash; both in terms of evaluation (and re-evaluation) and in terms of ordering the statements properly.&nbsp; This innovative approach for inference allows savings in both areas.&nbsp; Let&rsquo;s explore how.</p>
<p>In absence of this algorithm, or derivatives, the traditional approach is to sequence and nest the&nbsp;IF-THEN-ELSE statements of the rules in such a way that you capitalize on the test evaluations &mdash; you reuse them as much as possible.&nbsp; When you think about it, it would be akin to creating the leanest decision tree possible, in a&nbsp;graphical form or more likely in straight code.&nbsp; This &ldquo;sequential&rdquo; approach breaks (mostly in terms of performance) when rules need to execute as a result of the execution of other rules.&nbsp; This is what we call inference.&nbsp; In a sequential approach, you would have to restart the complete ruleset evaluation after each rule execution (if rules are prioritized) or after a full pass through the ruleset (in absence of&nbsp;prioritization).&nbsp; Keep in mind too that prioritization might prevent you from reusing test evaluations all together in a sequential approach: if the conditions are not co-located in the logic, then you will likely have to perform those tests multiple times.&nbsp; Systems have become much more sophisticated than they used to be back then, and most BRMS products provide a sequential deployment option which is still pretty good when inference is not needed.&nbsp; The beauty of those products is that you can still manage the rules independently of each other and the sequential engine takes care of optimizing the generated spaghetti code.</p>
<p>For illustration, I&rsquo;ll use a simple set of rules that we are mostly familiar with: airline miles calculations.&nbsp; The decision service is responsible for processing the frequent flyer&rsquo;s activity.</p>
<blockquote>
<p>if award miles for last year or current year &gt; 25,000 then status = Silver</p>
<p>if award miles for last year or current year &gt; 100,000 then status = Gold</p>
<p>If flight is less than 500 miles then award 500 miles</p>
<p>If flight is 500 miles or more then award flight miles</p>
<p>if category is business or first then award 50% bonus miles</p>
<p>if status is Gold and airline is not partner then award 100% bonus miles</p>
<p>if status is Silver&nbsp;and airline is not partner then award 20% bonus miles</p>
<p>if member signed up for 3-flights-for-5k-in-March and number of return flights in March 2011 = 3 then award 5,000 additional miles</p>
<p>if status becomes Gold then award 8 upgrade certificates</p>
</blockquote>
<p>I will live it up to the creative minds of the Airline marketers to add more promotions to our short list here: hotel or rental car bonuses, incentive programs, lifetime members, accelerator programs, etc.</p>
<p>This example does not use a lot of inference but that is enough to illustrate our point.&nbsp; If you run through the rules, sequentially &mdash; let&rsquo;s keep the current order &mdash; and this&nbsp;Frequent Flyer member reaches Gold status in that very transaction, we need to process the rules once more to update the status and award the proper bonus miles and other incentives.</p>
<p>Inference engines assess all the applicable rules and then fire the first applicable rule, propagate the changes and fire again.&nbsp; How does that work?</p>
<p><strong>Constructing the Rete Network</strong></p>
<p>The Rete network is the heart of the Rete algorithm.&nbsp; It is made of nodes that each hold a list of objects that satisfy the associated condition.&nbsp; The original Rete algorithm worked out of facts but I will simplify the description to refer to objects since all commercial engines have evolved to be object-oriented nowadays.</p>
<p>The Rete network is constructed when you compile a rules project &mdash; only once when you start that service (or class for simpler designs) &mdash;&nbsp;and then shared across all invocations.</p>
<p>The discrimination tree is the first part of the Rete network.&nbsp; It starts with Alpha nodes that are associated with Classes (in the object-oriented sense).&nbsp; All instances of a given class will be listed in any given Alpha node.&nbsp; The discrimination happens by adding conditions as single nodes attached to the Alpha node or to another parent node.</p>
<p>Let&rsquo;s review what that means for our Airline example:</p>
<p>Alpha nodes are created for each class: Frequent Flyer Account and Flight.</p>
<p><img src="https://www.sparklinglogic.com/wp-content/uploads/2011/03/rete-alpha-nodes-1024x267.jpg" alt="Rete - Alpha Nodes"></p>
<p>Conditions are then appended:</p>
<ul>
<li>Frequent Flyer
<ul>
<li>status is Gold</li>
<li>status is Silver</li>
<li>award miles for last year or current year &gt; 25k</li>
<li>award miles for last year or current year &gt; 100k</li>
<li>signed up for &ldquo;3-flights-for-5k-in-March&rdquo; promotion
<ul>
<li>number of flights in March = 3</li>
</ul>
</li>
</ul>
</li>
<li>Flight
<ul>
<li>miles &gt;= 500</li>
<li>miles &lt; 500</li>
<li>airline is not a partner</li>
<li>airline is partner</li>
<li>category is business or first</li>
</ul>
</li>
</ul>
<p>etc.</p>
<p>Each node represents an additional test to the series of conditions applied upstream.&nbsp; If you follow a path from top to bottom, you should be able to read the complete set of conditions THAT APPLY TO ONE GIVEN CLASS OF OBJECTS.</p>
<p><img src="https://www.sparklinglogic.com/wp-content/uploads/2011/03/rete-discrimination-tree1-1024x373.jpg" alt="Rete - Discrimination Tree"></p>
<p>Finally the nodes are connected across classes.&nbsp; This is where we can combine the conditions for a non-partner flight taken by a Gold member.&nbsp; We call those &ldquo;joints&rdquo; as we will combine the list of objects that verify conditions on one branch with the list of objects that verify the conditions on another branch.</p>
<p><img src="https://www.sparklinglogic.com/wp-content/uploads/2011/03/rete-network-1024x656.jpg" alt="Rete Network"></p>
<p>On the performance side, you may have been warned in the past not to combine too many patterns in a single rule.&nbsp; It is clear looking under the hood that the product of all possible accounts by all possible flights could result into a combinatorial explosion. The more discrimination upfront, the better obviously.</p>
<p>The path eventually end with the action part of the rules.&nbsp; The content of the actions is irrelevant for the Rete network.&nbsp; you could replace the labels here with the name of the rule (I did not name the rules in our example so I displayed the actual actions).&nbsp; The rules can be reconstituted by following the incoming paths.&nbsp; All nodes are AND-ed.&nbsp; This is an interesting point. ORs are typically not natively supported: rules are duplicated for each OR-ed flavor.&nbsp; If you were to add a rule to grant the same bonus for SILVER or GOLD customers traveling to Hawaii then you would just have nodes connected to the existing GOLD and SILVER nodes as if they were written as separate rules.&nbsp; In terms of performance or maintenance, it does not matter though since the Rete algorithm handles the supposed duplication as efficiently as any manual code would.</p>
<p>It is a great time to emphasize the great savings you would get when the number of rules increases.&nbsp; Let&rsquo;s say that you want to introduce a new promotion for GOLD customers to/from specific destinations or for reached milestones (50k and 75k award miles).&nbsp; The nodes that test for GOLD status do not need to be created or duplicated.&nbsp; The test as well as the execution of the test is leveraged and reused transparently regardless of the order in which the rules need to be executed.</p>
<p><strong>Rete Cycle: Evaluate</strong></p>
<p>Let&rsquo;s look now at what happens at runtime when you invoke that service with customer data.&nbsp; Your design will dictate whether those transactions are applied &ldquo;real-time&rdquo; (or on-demand) or whether you want to execute a batch of transactions at once, at night when you have access to your Mainframe window for example.&nbsp; Rules can actually be used either way and in many projects, they are actually used both ways by two different systems.&nbsp; For an Insurance underwriting project, you might deploy the same rules online in your self-service website as well as your older batch application that processes the applications coming from your brokers.</p>
<p>Let&rsquo;s assume that we are processing transactions one by one for simplicity.</p>
<p>The evaluation phase consist in running the data through the Rete network to identify the applicable rules, for which all conditions are satisfied.&nbsp; The nodes in the network will hold lists of objects that satisfy the conditions in the incoming path.</p>
<p>In our airline example, Joe flies from Washington DC (IAD) to San Francisco (SFO).&nbsp; Flight accounts for 2,419 miles I recall.&nbsp; Starting with already 150k award miles, Joe should bank double miles (4,838) for his GOLD status.&nbsp; How does Rete propagates those facts to make the same decision?</p>
<p><img src="https://www.sparklinglogic.com/wp-content/uploads/2011/03/rete-propagate1-1024x708.jpg" alt="Rete - Propagate"></p>
<p>The Account Alpha node references Joe as a Frequent Flyer.&nbsp; Because of his 150k award miles, he also qualifies for the &ldquo;&gt; 100k&rdquo; condition.&nbsp; He is referenced in this node too.&nbsp; My handwriting did not allow me to write down Joe in the small real estate so I just highlighted the node with a X.&nbsp; In reality, Joe would be referenced specifically as we need to keep track of who qualifies in case we process multiple objects at once.&nbsp; For this exercise, I also assume that service does not store the GOLD status and computes it every time so the associated node still has an empty list.</p>
<p>Similarly, the IAD-SFO flight is referenced in the list of all flights in the transaction.&nbsp; The flight is more than 500 miles and on the airline.&nbsp; The associated lists reference the flight accordingly.</p>
<p>At that point in time we do not have any joint to worry about since the account status is not known yet to qualify for GOLD.</p>
<p>2 rules are satisfied.&nbsp; All facts have been propagated.&nbsp; We are ready for execution.</p>
<p><strong>Rete Cycle:</strong> <strong>Execute</strong></p>
<p>The rules for which all conditions are satisfied are said to be active in the agenda.&nbsp; The agenda contains the list of all rules that should be executed, along with the list of objects that are responsible for the conditions to be true.&nbsp; I insist on the word &ldquo;SHOULD&rdquo;.&nbsp; The agenda will sort them according to their priorities (and other conflict resolution methods).&nbsp; If the execution of a rules invalidates a rule with a lower priority, then this second rule will logically never execute.&nbsp; This would have been the case if Joe had started with 99k award miles &ndash; he would have qualified for the SILVER bonus until we bank the IAD-SFO miles and granted him GOLD status.&nbsp; Joe can qualify for SILVER or GOLD bonus miles but certainly not both.</p>
<p>Priorities are a touchy subject in the Rules community as some purists are set against its use.</p>
<p>The agenda lists the 2 rules (GOLD status and 500+ Flight Miles).&nbsp; The first one is executed.&nbsp; In our airline example, order does not matter for those 2 rules.&nbsp; Let&rsquo;s assume that flight miles are awarded first.&nbsp; Joe gets 152,419 miles in his account so far.</p>
<p><strong>Do it again and again&hellip;</strong></p>
<p>After the execution of the first rule, facts are propagated again.&nbsp; Flights have not changed so none of that subtree is re-evaluated.&nbsp; Joe&rsquo;s account has changed so the conditions related to award miles are re-assessed, with no change in our example.&nbsp; The GOLD status rule remains active in the agenda.</p>
<p>We are ready to execute the following rule: GOLD Status which is the only rule left in the agenda.</p>
<p>As a result, the GOLD Status node is updated for Joe, which leads to the GOLD status and not partner node to update as well, which leads to the 100% GOLD bonus rule to become activated in the agenda.</p>
<p><img src="https://www.sparklinglogic.com/wp-content/uploads/2011/03/rete-execute-and-propagate1-1024x728.jpg" alt="Rete - Execute and Propagate"></p>
<p>In the end, Joe gets the proper credit of 154,839 in his award miles account.</p>
<p>As you add more rules that apply to GOLD or SILVER customers, the RETE network will grow always reusing the same nodes as much as possible &mdash; which is the part that takes time to do properly by hand if you are still coding it manually.&nbsp; Those spaghetti in the &ldquo;Joint&rdquo; part of the algorithm are a nightmare to maintain properly and can cost you significant debug time.&nbsp; This trivial example can become quickly messy when you consider the many different ways you can qualify for status with eligible segments rather than eligible miles, with accelerators, with lifetime miles, as well as the incentive promotions that you might get by default and those that you need to sign up for.&nbsp; Reusing those conditions &mdash; and their execution &mdash; regardless of when and where they show up is a performance boost that largely compensate for the initial overhead when inference is needed.</p>
<p>When inference is not needed and you only go through one cycle of propagation, it is debatable of course.</p>
<p>In my opinion, Charles opened the door for an algorithmic solution to properly ensuring the enforcement of many rules in an uncertain order.&nbsp; In other words, he created an approach where very large volumes of rules could be assessed in a short time &mdash; much faster than any other algorithms when rules need to be executed repeatedly on a given data set. The further performance improvements he introduced in Rete II, Rete III and lately in Rete-NT have also changed the initial &ldquo;ideal&rdquo; characteristics of a Rete project to better support much larger data set &mdash; what was marketed as the Rete wall by the non-Rete vendors.</p>
<p>It is somewhat surprising that no new algorithm break-through (other than Charles&rsquo;s) were made since then.&nbsp; The basis for Business Rules execution is vastly based on Rete, or sequential, but hardly anything else.&nbsp; It reminds me of the sharks and their evolution.&nbsp; I remember, in my life science days, learning that sharks have not evolved at all in a very long time&hellip;&nbsp;likely because they are perfectly adapted to their environment.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Sparkling Logic https://ift.tt/2TArJfj<br>
<br>
March 14, 2019 at 05:46PM
